### DISCUSSION TITLE
MicroPython imports a binary format class library through the `import` directive?

### DISCUSSION THREAD

**Discussion by Mars-CN** [micropython/micropython#14481] [Core Development] [answered]
MicroPython imports a binary format class library through the `import` directive? Just like Python, the class library file (mylib.so) that uses C/C++ extension can be used directly after importing it through `import mylib`, but MicroPython can only import files in .py or .mpy format. I don’t know if it can Import files in so format?
Or, where should I start modifying the source code?
PS: This class library is loaded dynamically, and I don’t want to compile it into MicroPython’s firmware.

MicroPython 通过 `import`指令导入一个二进制格式的类库？就像Python 一样，使用C/C++ 扩展的类库文件(mylib.so) 通过 `import mylib` 导入后就可以直接使用了，但是 MicroPython只能导入 .py或者.mpy 格式的文件，不知道是否能导入 so 格式的文件？
或者，我应该从哪里入手修改源代码？
PS：这个类库是动态化加载进去的，我不想将它编译到 MicroPython的固件中。

**Comment by peterhinch** [comment]
CPython does this via [ctypes](https://stackoverflow.com/questions/48919761/how-to-run-so-files-using-through-python-script). The equivalent library in MicroPython is [uctypes](https://docs.micropython.org/en/latest/library/uctypes.html). Unfortunately this does not have `cdll` or `CDLL` objects so it seems that using `.so` files is not possible.

**Comment by jonnor** [comment, marked as answer]
If you have access to the C/C++ code, you can build a dynamic native module. This can be built separately from main firmware, copied to device and loaded at runtime. See the documentation https://docs.micropython.org/en/latest/develop/natmod.html#supported-features-and-limitations and the examples https://github.com/micropython/micropython/tree/master/examples/natmod

**Reply by Mars-CN** [reply]
Thank you very, very, very much!!!

**Reply by Mars-CN** [reply]
@jonnor 
I encountered some issues because I needed to use the content of string. h in the code, but there was a prompt during compilation
`LinkError: build/mpfile.o: undefined symbol: memset A type of error`
I also encountered the same error when using some libraries in MicroPython
`LinkError: build/mpfile.o: undefined symbol: mp_vfs_listdir`

How should these be resolved?
Do I feel that it is related to static class libraries
